File: leetcode/Python/maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.allSum = 0
        self.diff = 1000000007
        self.MOD = 1000000007
        self.a = 0
        self.b = 0
        self.t = 0
    def traverseSum(self, root):
        sol = 0
        q = collections.deque()
        q.append(root)
        while q:
            t = q.popleft()
            if t.left:
                q.append(t.left)
            if t.right:
                q.append(t.right)
            sol+=t.val
        return sol
    def findNum(self, node):
        if not node:
            return 0
        l=self.findNum(node.left)
        r=self.findNum(node.right)
        self.t=max(self.allSum-l,l)-min(self.allSum-l,l)
        if self.diff>self.t:
            self.diff, self.a, self.b = self.t, self.allSum-l, l
        self.t=max(self.allSum-r,r)-min(self.allSum-r,r)
        if self.diff>self.t:
            self.diff, self.a, self.b = self.t, self.allSum-r, r
        return l+r+node.val
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        self.allSum = self.traverseSum(root)
        self.findNum(root)
        sol = (self.a*self.b)%self.MOD
        return sol

File: leetcode/C++/maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diff, allSum, a, b, t;
    const int MOD=1000000007;
    void traverseSum(TreeNode* node){
        TreeNode* n;
        queue<TreeNode*> q;
        q.push(node);
        while(!q.empty()){
            n=q.front(); q.pop();
            if(n->left) q.push(n->left);
            if(n->right) q.push(n->right);
            allSum+=n->val;
        }
    }
    int calcMax(TreeNode* root){
        if(root==nullptr) return 0;
        int left=calcMax(root->left);
        int right=calcMax(root->right);
        t = max(allSum-left,left)-min(allSum-left,left);
        if(diff>t) {diff=t; a=allSum-left; b=left;}
        t = max(allSum-right,right)-min(allSum-right,right);
        if(diff>t) {diff=t; a=allSum-right; b=right;}
        return left+right+root->val;
    }
    int maxProduct(TreeNode* root) {
        long long sol = 0;
        diff=MOD;
        allSum=a=b=0;
        traverseSum(root);
        calcMax(root);
        sol=(long long)(a)*(b)%MOD;
        return sol;
    }
};